File: ScaleformUI_V8/src/menus/UIMenu/windows/uimenudetailswindow.ts
import { Vector2 } from "math/vector2";
import { UIMenuWindow } from "./uimenuwindow";
import { SColor } from "elements/scolor";
import { ScaleformUI } from "scaleforms/scaleformui/main";

export class UIDetailImage {
    public Txd: string
    public Txn: string
    public Pos: Vector2
    public Size: Vector2

    constructor(txd: string, txn: string, pos: Vector2, size: Vector2) {
        this.Txd = txd;
        this.Txn = txn;
        this.Pos = pos;
        this.Size = size;
    }
}

export class UIDetailStat {
    public Percentage: number;
    public HudColor: SColor;
    constructor(percentage: number, color: SColor) {
        this.Percentage = percentage;
        this.HudColor = color;
    }
}


export class UIMenuDetailsWindow extends UIMenuWindow {
    public DetailTop: string;
    public DetailMid: string;
    public DetailBottom: string;
    public DetailLeft: UIDetailImage;
    public StatWheelEnabled: boolean;
    public DetailStats: UIDetailStat[];

    constructor(...args: any[]) {
        super();
        this.id = 1;
        this.DetailTop = args[0];
        this.DetailMid = args[1];
        this.DetailBottom = args[2];
        this.StatWheelEnabled = false;
        this.DetailLeft = new UIDetailImage("", "", Vector2.zero, Vector2.zero);
        this.DetailStats = [];

        if (args.length === 3 || args.length === 4) {
            if (args.length === 4) {
                this.DetailLeft = args[3] || this.DetailLeft;
            }
        } else if (args.length === 5) {
            this.StatWheelEnabled = args[3];
            this.DetailStats = args[4] as UIDetailStat[];
            this.DetailLeft = this.DetailLeft;
        }
    }

    public UpdateLabels(top: string, mid: string, bot: string, leftDetail: UIDetailImage) {
        this.DetailTop = top;
        this.DetailMid = mid;
        this.DetailBottom = bot;
        this.DetailLeft = leftDetail == undefined ? new UIDetailImage("", "", Vector2.zero, Vector2.zero) : leftDetail;
        if (this.ParentMenu != undefined && this.ParentMenu.Visible) {
            let wid = this.ParentMenu.Windows.indexOf(this);
            if (!this.StatWheelEnabled)
                ScaleformUI.Scaleforms._ui?.callFunction("UPDATE_DETAILS_WINDOW_VALUES", wid, this.DetailBottom, this.DetailMid, this.DetailTop, this.DetailLeft.Txd, this.DetailLeft.Txn, this.DetailLeft.Pos.x, this.DetailLeft.Pos.y, this.DetailLeft.Size.x, this.DetailLeft.Size.y);
            else
                ScaleformUI.Scaleforms._ui?.callFunction("UPDATE_DETAILS_WINDOW_VALUES", wid, this.DetailBottom, this.DetailMid, this.DetailTop, "statWheel");
        }
    }

    public AddStatsListToWheel(stats: UIDetailStat[]) {
        if (this.StatWheelEnabled) {
            this.DetailStats = stats;
            if (this.ParentMenu != undefined && this.ParentMenu.Visible) {
                let wid = this.ParentMenu.Windows.indexOf(this);
                stats.forEach((value: UIDetailStat) => {
                    ScaleformUI.Scaleforms._ui?.callFunction("ADD_STATS_DETAILS_WINDOW_STATWHEEL", wid, value.Percentage, value.HudColor);
                });
            }
        }
    }

    public AddStatSingleToWheel(stat: UIDetailStat) {
        if (this.StatWheelEnabled) {
            this.DetailStats.push(stat);
            if (this.ParentMenu != undefined && this.ParentMenu.Visible) {
                let wid = this.ParentMenu.Windows.indexOf(this);
                ScaleformUI.Scaleforms._ui?.callFunction("ADD_STATS_DETAILS_WINDOW_STATWHEEL", wid, stat.Percentage, stat.HudColor);
            }
        }
    }

    public UpdateStatsToWheel() {
        this.UpdateStatsToWheel(this.DetailStats);
    }
    public UpdateStatsToWheel(stats: UIDetailStat[]) {
        if (this.StatWheelEnabled) {
            if (this.DetailStats.length != stats.length) {
                throw new Error("You cannot add items using this function");
            }
            this.DetailStats = stats;
            if (this.ParentMenu != undefined && this.ParentMenu.Visible) {
                let wid = this.ParentMenu.Windows.indexOf(this);
                stats.forEach((value: UIDetailStat) => {
                    ScaleformUI.Scaleforms._ui?.callFunction("UPDATE_STATS_DETAILS_WINDOW_STATWHEEL", wid, this.DetailStats.indexOf(value), value.Percentage, value.HudColor);
                });
            }
        }
    }

    public RemoveStatToWheel(stat: UIDetailStat) {
        let index = this.DetailStats.findIndex(s => s.Percentage === stat.Percentage && s.HudColor === stat.HudColor);
        if (index !== -1) {
            this.DetailStats.splice(index, 1);
        }
    }
    public RemoveStatToWheel(idx: number) {
        if (idx < 0 || idx >= this.DetailStats.length) return;
        this.DetailStats.splice(idx, 1);
        if (this.ParentMenu != undefined && this.ParentMenu.Visible) {
            let wid = this.ParentMenu.Windows.indexOf(this);
            ScaleformUI.Scaleforms._ui?.callFunction("REMOVE_STATS_DETAILS_WINDOW_STATWHEEL", wid, idx);
        }
    }
}

File: ScaleformUI_Csharp/Menus/UIMenu/Windows/UIMenuDetailsWindow.cs
using ScaleformUI.Elements;
using System.Drawing;

namespace ScaleformUI.Menu
{
    public class UIMenuDetailsWindow : UIMenuWindow
    {
        public string DetailTop;
        public string DetailMid;
        public string DetailBottom;
        public UIDetailImage DetailLeft;
        public bool StatWheelEnabled;
        public List<UIDetailStat> DetailStats;
        public UIMenuDetailsWindow(string top, string mid, string bot, UIDetailImage leftDetail = null)
        {
            id = 1;
            DetailTop = top;
            DetailMid = mid;
            DetailBottom = bot;
            StatWheelEnabled = false;
            DetailLeft = leftDetail ?? new()
            {
                Txd = "statWheel"
            };
        }

        public UIMenuDetailsWindow(string top, string mid, string bot, bool statWheelEnabled, List<UIDetailStat> details)
        {
            id = 1;
            DetailTop = top;
            DetailMid = mid;
            DetailBottom = bot;
            StatWheelEnabled = statWheelEnabled;
            DetailStats = details;
            DetailLeft = new()
            {
                Txd = "statWheel"
            };
        }

        public void UpdateLabels(string top, string mid, string bot, UIDetailImage leftDetail = null)
        {
            DetailTop = top;
            DetailMid = mid;
            DetailBottom = bot;
            DetailLeft = leftDetail ?? new()
            {
                Txd = "statWheel"
            };
            ParentMenu?.SetWindows(true);
        }

        public void AddStatsListToWheel(List<UIDetailStat> stats)
        {
            if (StatWheelEnabled)
            {
                DetailStats = stats;
                ParentMenu?.SetWindows(true);
            }
        }

        public void AddStatSingleToWheel(UIDetailStat stat)
        {
            if (StatWheelEnabled)
            {
                DetailStats.Add(stat);
                ParentMenu?.SetWindows(true);
            }
        }

        public void UpdateStatsToWheel()
        {
            UpdateStatsToWheel(DetailStats);
        }
        public void UpdateStatsToWheel(List<UIDetailStat> stats)
        {
            if (StatWheelEnabled)
            {
                if (DetailStats.Count != stats.Count)
                {
                    throw new Exception("You cannot add items using this function");
                }
                DetailStats = stats;
                ParentMenu?.SetWindows(true);
            }
        }
        public void RemoveStatToWheel(UIDetailStat stat)
        {
            UIDetailStat s = DetailStats.FirstOrDefault(x => x.Percentage == stat.Percentage && x.HudColor == stat.HudColor);
            if (s != null)
            {
                RemoveStatToWheel(DetailStats.IndexOf(s));
            }
        }
        public void RemoveStatToWheel(int id)
        {
            if (id < 0 || id >= DetailStats.Count) return;
            DetailStats.RemoveAt(id);
            ParentMenu?.SetWindows(true);
        }
    }

    public class UIDetailImage
    {
        public string Txd { get; set; }
        public string Txn { get; set; }
        public PointF Pos { get; set; }
        public SizeF Size { get; set; }

        public UIDetailImage()
        {
            Txd = "";
            Txn = "";
            Pos = new(0, 0);
            Size = new(0, 0);
        }
        public UIDetailImage(string txd, string txn, PointF pos, SizeF size)
        {
            Txd = txd;
            Txn = txn;
            Pos = pos;
            Size = size;
        }
    }

    public class UIDetailStat
    {
        public int Percentage;
        public SColor HudColor;
        public UIDetailStat(int percentage, SColor color)
        {
            Percentage = percentage;
            HudColor = color;
        }
    }
}
